libraries/chain/include/eosio/chain/finality/qc.hpp

```diff
@@ -0,0 +1,261 @@
+
+      bool is_strong() const {
+         return active_policy_sig.is_strong() && (!pending_policy_sig || pending_policy_sig->is_strong());
+      }
+
+      bool is_weak() const {
+         return active_policy_sig.is_weak() || (pending_policy_sig && pending_policy_sig->is_weak());
+      }
+
+      qc_claim_t to_qc_claim() const {
+         return { .block_num = block_num, .is_strong_qc = is_strong() };
+      }
+   };
+
+   struct qc_data_t {
+      std::optional<qc_t> qc;  // Comes either from traversing branch from parent and calling get_best_qc()
+                               // or from an incoming block extension.
+      qc_claim_t qc_claim;     // describes the above qc_t. In rare cases (bootstrap, starting from snapshot,
+                               // disaster recovery), we may not have a qc_t so we use the `lib` block_num
+                               // and specify `weak`.
+   };
+
+   /**
+    * All public methods are thread-safe.
+    * Used for incorporating votes into a qc signature.
+    * "open" in that it allows new votes to be added at any time.
+    */
+   class open_qc_sig_t {
+   public:
+      enum class state_t {
+         unrestricted,  // No quorum reached yet, still possible to achieve any state.
+         restricted,    // Enough `weak` votes received to know it is impossible to reach the `strong` state.
+         weak_achieved, // Enough `weak` + `strong` votes for a valid `weak` QC, still possible to reach the `strong` state.
+         weak_final,    // Enough `weak` + `strong` votes for a valid `weak` QC, `strong` not possible anymore.
+         strong         // Enough `strong` votes to have a valid `strong` QC
+      };
+
+      struct votes_t : fc::reflect_init {
+      private:
+         friend struct fc::reflector<votes_t>;
+         friend struct fc::reflector_init_visitor<votes_t>;
+         friend struct fc::has_reflector_init<votes_t>;
+         friend class open_qc_sig_t;
+
+         vote_bitset                    bitset;
+         bls_aggregate_signature        sig;
+         struct bit_processed {
+            alignas(hardware_destructive_interference_size)
+            std::atomic<bool> value;
+         };
+         std::vector<bit_processed> processed; // avoid locking mutex for _bitset duplicate check
+
+         void reflector_init();
+      public:
+         explicit votes_t(size_t num_finalizers)
+            : bitset(num_finalizers)
+            , processed(num_finalizers) {}
+
+         // thread safe
+         bool has_voted(size_t index) const;
+
+         vote_status add_vote(size_t index, const bls_signature& sig);
+
+         template<class CB>
+         void visit_bitset(const CB& cb) const {
+            for (size_t i = 0; i < bitset.size(); ++i) {
+               if (bitset[i])
+                  cb(i);
+            }
+         }
+      };
+
+      open_qc_sig_t();
+
+      explicit open_qc_sig_t(size_t num_finalizers, uint64_t quorum, uint64_t max_weak_sum_before_weak_final);
```

Comment:
Why not just have a constructor taking a `const finalizer_policy_ptr&` to avoid duplicated code at lines 218-226